src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1757 @@
+            ), "Make sure that config.axial_pos_shape factors: {} multiply at least to max(sequence_length, least_common_mult_chunk_length): max({}, {})".format(
+                self.axial_pos_shape, sequence_length, self.least_common_mult_chunk_length,
+            )
+
+            # reshape axial encodings and use only until sequence_length
+            position_encodings = torch.cat(broadcasted_weights, dim=-1)
+            position_encodings = position_encodings.view(batch_size, -1, position_encodings.shape[-1])[
+                :, :sequence_length
+            ]
+
+        return position_encodings
+
+
+class PositionEmbeddings(nn.Module):
+    """Constructs conventional position embeddings of shape `[max_pos_embeddings, hidden_size]`.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.dropout = config.hidden_dropout_prob
+        self.embedding = nn.Embedding(config.max_position_embeddings, config.hidden_size)
+
+    def forward(self, position_ids):
+        position_embeddings = self.embedding(position_ids)
+        position_embeddings = nn.functional.dropout(position_embeddings, self.dropout, self.training)
+        return position_embeddings
+
+
+class ReformerEmbeddings(nn.Module):
+    """Construct the embeddings from word, position and token_type embeddings.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.max_position_embeddings = config.max_position_embeddings
+        self.dropout = config.hidden_dropout_prob
+
+        self.word_embeddings = nn.Embedding(config.vocab_size, config.hidden_size)
+        self.position_embeddings = (
+            AxialPositionEmbeddings(config) if config.axial_pos_embds else PositionEmbeddings(config)
+        )
+
+    def forward(self, input_ids=None, position_ids=None, inputs_embeds=None):
+        if input_ids is not None:
+            input_shape = input_ids.size()
+            device = input_ids.device
+        else:
+            input_shape = inputs_embeds.size()[:-1]
+            device = inputs_embeds.device
+
+        seq_length = input_shape[1]
+        if position_ids is None:
+            position_ids = torch.arange(seq_length, dtype=torch.long, device=device)
+            position_ids = position_ids.unsqueeze(0).expand(input_shape)
+
+        if inputs_embeds is None:
+            inputs_embeds = self.word_embeddings(input_ids)
+
+        assert (
+            position_ids.shape[-1] <= self.max_position_embeddings
+        ), "Sequence Length: {} has to be larger equal than config.max_position_embeddings: {}".format(
+            position_ids.shape[-1], self.max_position_embeddings
+        )
+
+        embeddings = nn.functional.dropout(inputs_embeds, self.dropout, self.training)
+        position_embeddings = self.position_embeddings(position_ids)
+
+        embeddings = embeddings + position_embeddings
+        embeddings = nn.functional.dropout(embeddings, self.dropout, self.training)
```

Comment:
dropout on each individually and their sum?
Maybe just have this class hold all the dropout logic?

Reply:
Great catch! I checked the original code again and I had a dropout too much there. Corrected it :-)